Initial prototype implementation of an email scrapper.

scande3 opened the pull request:
See README for relevant usage notes and implementation concerns.

Commit:
Initial prototype implementation of an email scrapper

johnfertitta commented on scrap_emails.py:
```diff
@@ -0,0 +1,125 @@
+        # Ensure that we don't parse things like images or pdfs
+        if guessed_type[0] is None or guessed_type[0].startswith('text'):
+            # Ensure that we stick only to this domain.
+            # FIXME: There is an issue with http vs https in this sample application. Also with www vs non-www.
```
How do you think you could fix this?

scande3 replied:
Instead of comparing against the full base url, I've stripped out the "http(s)://" portion of it. Visited urls also only store the http-less portion of their url now which has the added benefit of saving on memory space.

johnfertitta commented on scrap_emails.py:
```diff
@@ -0,0 +1,125 @@
+            # FIXME: There is an issue with http vs https in this sample application. Also with www vs non-www.
+            if l.startswith(tuple(self.start_urls)):
+                # Exclude sites we have already visited
+                if l not in self.existing_visited:
```
This check could get slow with large websites, is there a better way we can be storing visited urls?

scande3 replied:
I should have used sets over arrays since the values are unique. The set lookup is O(1) compared to the O(n) of an array comparison.

There are a few other things that could be done as well. For example to use less memory at the cost of some CPU cycles, one could md5 hash the url for comparison purposes. That would give a consistent 32 characters in memory compared to the [average ~77 characters of
a url](http://www.supermind.org/blog/740/average-length-of-a-url-part-2). One could even do a mixture of hashing and not hashing depending on if the url is greater than 32 characters or not.

Furthermore, if one was trying to get all emails from an exceptionally large site where memory usage of the set was a concern, one of two approaches is possible:

1. Write each email or url output to an indexed database field or the an indexer like Solr. This isn't a fast operation but would eliminate the need for a large in-memory set.

2. Split up the subdirectory parsing.
    1. From the index page, create a set of all subdirectories that exist (ie. one for web.mit.edu/faculty, one for web.mit.edu/industry, etc). 
    2. Then process each of those subdirectories individually forcing a match on that base url with that subdirectory. 
    3. If you encounter a new subdirectory that wasn't on the index page, then add it to the list subdirectories to process.
    4. Only process urls that exist in the current subdirectory path.
    5. Once a subdirectory is complete, remove this subdirectory from the "to process" list and clear the set of existing visited urls.
    6. Repeat the process for the next subdirectory.

2b. This second approach essentially splits the site up into smaller site pieces. It does have the disadvantage of potentially missing some emails though. (For example, a link on "faculty"
points to a page on "industry" that isn't linked to from within the "industry" subdirectory. This should be a fairly rare case though).

johnfertitta commented on scrap_emails.py:
```diff
@@ -0,0 +1,125 @@
+        # Final check to not at some numbers that were matching the pattern.
+        return_emails = []
+        for email in re.findall(self.regex, s):
+            if email[0] not in self.matched_emails:
```
This could be slow as well, any other way we can store these?

scande3 replied:
Same as previous reply on this topic.

johnfertitta commented on scrap_emails.py:
```diff
@@ -0,0 +1,125 @@
+                if not email[0].startswith('//'):
+                    if  mimetypes.guess_type(email[0])[0] is None or (not mimetypes.guess_type(email[0])[0].startswith('image') and not mimetypes.guess_type(email[0])[0].startswith('video')):
+                        cleaned_email = re.sub(r'\sat\s', '@', email[0])
+                        if not re.match(r'.+@[\d\.]+$', cleaned_email):
```
This nesting gets pretty deep, can it be refactored to be a bit easier to read and follow?

scande3 replied:
I've split this up into two methods:

1. A method to clean up a matched email so that it is standardized in a "<email>@<domain>" format.

2. A method for determining if the regex matched email was a correctly matched email.

This reduces the nested logic in that one method. Additionally, it makes it easier to expand the code to include more variations one might try to write their email as on a website.

Not sure if there is more I should do to make it more readable?

johnfertitta commented on scrap_emails.py:
```diff
@@ -0,0 +1,125 @@
+    # Don't parse more than 4 levels deep for this example.
+    # TODO: Make this configurable
+    custom_settings = {
+        'DEPTH_LIMIT': 4
```
Good job limiting the depth

Commit:
Updates from code by @johnfertitta for previous commit (fce6c4f)

I made a few other small changes beyond those listed in these code review comments. Specifically:

1. Standardizing the email before adding it to the set of existing emails. This fixed having unintentional duplicates in that list if they were listed in different formats throughout the site. (ie. email at domain.com and email@domain.com).

2. Changed the main regex to require at least two characters after the end ".". This was to fix a case of name@c.e.r.n that was invalid.

3. Replacing " dot " with "." and removing "%20" at the start of emails.

4. Not a change but a left-in commented bit of code. If you want the output to include the link where the email came from (which I used in debugging), uncomment out: yield {'email': email, 'link': response.url}. Then comment out the line above that to be: #yield {'email': email}.

A list of fixes from the code review:

1. Changed arrays to sets for validating already processed links and emails.

2. Split of the email parsing logic into seperate methods for better readability and future code expansion.

3. Elimate "http(s)://" from url matching to allow for http and https urls to work the same and reduce memory footprint of the set.

scande3 commented on scrap_emails.py:
```diff
@@ -0,0 +1,125 @@
+        # Ensure that we don't parse things like images or pdfs
+        if guessed_type[0] is None or guessed_type[0].startswith('text'):
+            # Ensure that we stick only to this domain.
+            # FIXME: There is an issue with http vs https in this sample application. Also with www vs non-www.
```
I should have used sets over arrays since the values are unique. The set lookup is O(1) compared to the O(n) of an array comparison.

There are a few other things that could be done as well. For example to use less memory at the cost of some CPU cycles, one could md5 hash the url for comparison purposes. That would give a consistent 32 characters in memory compared to the [average ~77 characters of
a url](http://www.supermind.org/blog/740/average-length-of-a-url-part-2). 

Furthermore, if one was trying to get all emails from an exceptionally large site where memory usage of the set was a concern, one of two approaches is possible:

1. Write each email or url output to an indexed database field or the an indexer like Solr. This isn't a fast operation but would eliminate the need for a large in-memory set.

2. Split up the subdirectory parsing.
    1. From the index page, create a set of all subdirectories that exist (ie. one for web.mit.edu/faculty, one for web.mit.edu/industry, etc). 
    2. Then process each of those subdirectories individually forcing a match on that base url with that subdirectory. 
    3. If you encounter a new subdirectory that wasn't on the index page, then add it to the list subdirectories to process.
    4. Only process urls that exist in the current subdirectory path.
    5. Once a subdirectory is complete, remove this subdirectory from the "to process" list and clear the set of existing visited urls.
    6. Repeat the process for the next subdirectory.

2b. This second approach essentially splits the site up into smaller site pieces. It does have the disadvantage of potentially missing some emails though. (For example, a link on "faculty"
points to a page on "industry" that isn't linked to from within the "industry" subdirectory. This should be a fairly rare case though).

scande3 commented on scrap_emails.py:
```diff
@@ -0,0 +1,125 @@
+            # FIXME: There is an issue with http vs https in this sample application. Also with www vs non-www.
+            if l.startswith(tuple(self.start_urls)):
+                # Exclude sites we have already visited
+                if l not in self.existing_visited:
```
Same as previous reply.

scande3 commented:
Should be all updated and there are notes in the commit message of the update at: https://github.com/scande3/kraken/pull/1/commits/c4f252aa90e9bebb9466b922840d47a1d5d76853

The most relevant is a way to include the url with the matched email that is useful for debugging and verification purposes.

Thanks for the review!
